### ISSUE TITLE
localSourcePath option doesn't seem to work

### ISSUE THREAD

**Issue by Petezah** [awslabs/aws-codebuild-jenkins-plugin#76]
I am currently trying to use the new "localSourcePath" option to try to speed up our parallel builds (multiple parallel codebuild builds which all use the same source), among other possible improvements once I get this working.  However, I consistently get a message like this:

[AWS CodeBuild Plugin] Local file to be uploaded to S3: /ebsvol/jenkins/workspace/my-workspace-name/src.zip
[AWS CodeBuild Plugin] /ebsvol/jenkins/workspace/my-workspace-name/src.zip (No such file or directory)

I create the zip file at the beginning with:

sh 'mkdir -p git'
dir('git') {
  checkout scm
}
sh 'rm -f src.zip'
zip dir: 'git', glob: '', zipFile: 'src.zip'

And then subsequently pass it into the awsCodeBuild plugin:

awsCodeBuild(
        projectName: codeBuildProjectName,
        buildSpecFile: buildSpecOverride,
        region: region, 
        credentialsId: credentialId, 
        credentialsType: 'jenkins', 
        sourceControlType: 'jenkins', 
        localSourcePath: 'src.zip',
        sseAlgorithm: 'AES256')

Am I doing something wrong here?  Or am I hitting some weird corner-case?

**Comment by leobaran-aws** [comment]
Can you run `sh 'pwd; ls'` after creating the zip file to ensure the zip is in `my-workspace-name/`?

**Comment by Petezah** [comment]
Yes, I tried that too.  It shows both the path it says is missing, and that the zip file is there (and the git repo subdirectory).

**Comment by leobaran-aws** [comment]
Can you send me information about your Jenkins setup so I can try to reproduce this?

**Comment by leobaran-aws** [comment]
Is it possible that the CodeBuild steps and zip steps are being run on different nodes?

**Comment by leobaran-aws** [comment]
Okay I think I was able to reproduce the issue. With an older Jenkins installation, the `zip` command failed and I had to install the [pipeline-utility-steps](https://plugins.jenkins.io/pipeline-utility-steps) plugin to get it to work. On a newer installation, the `zip` command silently failed, but once again worked after installing the pipeline-utility-steps plugin. Can you try installing that plugin and see if that fixes your build?

**Comment by Petezah** [comment]
Sorry for the delay--I've been working on about three different projects in the last day or so.  I did check, and we do have pipeline-utility-steps version 2.2.0, which looks to be the latest.  And we are running Jenkins 2.150.2 via the official Docker image.

Here is a simplified version of what I am doing in our Jenkinsfile:

```

def makeCodeBuildProject(
    Map params = [:],
    String codeBuildProjectName) {
    def buildSpecOverride = params["buildSpecOverride"] ?: ""
    def region = params["region"] ?: "us-east-2"
    def srcPath = params["sourcePath"] ?: ""

    def result = awsCodeBuild(
        projectName: codeBuildProjectName,
        buildSpecFile: buildSpecOverride,
        region: region, 
        credentialsId: 'my-credentials', 
        credentialsType: 'jenkins', 
        sourceControlType: 'jenkins', 
        localSourcePath: srcPath,
        sseAlgorithm: 'AES256')

    def artifactPath = result.getArtifactsLocation().split(':::')[1]
    return artifactPath
}

timestamps {
    def artifactPathSpecs = []
    node('master || ecs') {
        stage('Preparation') { // for display purposes
            sh 'mkdir -p git'
            dir('git') {
                // Get hub-core source
                checkout scm // get source from current branch/PR
            }

            sh 'rm -f src.zip'
            zip dir: 'git', glob: '', zipFile: 'src.zip'
        }
        stage('Build') {
            parallel "GCC": {
                sh 'pwd && ls'
                def artifactPath = makeCodeBuildProject(
                    region: "us-east-1",
                    sourcePath: "./src.zip",
                    'my-project-x64-use1-codebuild') //codeBuildProjectName
                artifactPathSpecs.push([s3path: artifactPath, localFolder: 'ubuntu-x64', region: 'us-east-1', friendlyName: 'GCC'])
            }, "Clang": {
                // Uses the same CodeBuild project (and ECR image) as ubuntu-x64 above
                def artifactPath = makeCodeBuildProject(
                    region: "us-east-1",
                    sourcePath: "./src.zip",
                    'my-project-x64-use1-codebuild', //codeBuildProjectName
                    buildSpecOverride: 'tools/docker_builds/ubuntu-x64/buildSpec-clang.yml')
                artifactPathSpecs.push([s3path: artifactPath, localFolder: 'ubuntu-x64-clang', region: 'us-east-1', friendlyName: 'Clang'])
            } //… other targets omitted ...

            failFast: false
        }
    }
}
```

In the "GCC" path, I was doing the `pwd && ls -la` as a sanity check, and it is outputting:

```
+ pwd
/ebsvol/jenkins/workspace/my-workspace-name
+ ls
git
git@tmp
src.zip
```

And this is the error output from the codebuild step:

```
[AWS CodeBuild Plugin] Using credentials provided by the DefaultAWSCredentialsProviderChain for authorization (provided from Jenkins credentials ****)
[AWS CodeBuild Plugin] Local file to be uploaded to S3: /ebsvol/jenkins/workspace/my-workspace-name/src.zip
[AWS CodeBuild Plugin] /ebsvol/jenkins/workspace/my-workspace-name/src.zip (No such file or directory)
```

It seems like the two things are happening on the same node, and the zip file appears to be there.

**Comment by leobaran-aws** [comment]
I wasn't able to reproduce the issue running a build on Jenkins 2.150.2 in a Docker image, although I was only using one master node. Can you tell me a bit more about your node setup?

**Comment by Petezah** [comment]
Ahh, okay.  This did occur to me last night too.  Yes, we have a more complicated setup.  We have a master instance in ECS which does almost no work, and these codebuild jobs are setup to occur on separate EC2 slave instances.  (Ignore the node name in my Jenkinsfile script--the ones designated to run the codebuild jobs are labeled "master", for... reasons... Our master instance is labeled "Jenkins".)

So I wonder if it is trying to read the zip file locally, instead of off of the agent's workspace?  That would explain why it shows up in the `ls`, but triggers a file not found.

**Comment by leobaran-aws** [comment]
I think I was able to reproduce the issue after setting up a node. We did [some work](https://github.com/awslabs/aws-codebuild-jenkins-plugin/commit/24337ae06bc161b425827bcc84b202bdc37afceb) a while back to properly handle remote workspaces, but it doesn't look like the new logic for local source is implemented improperly. I'll debug a bit and keep you updated.

**Comment by leobaran-aws** [comment]
Should be fixed by 29f2fa137e5b8c415e07c4d638dd56b7c9ccca54. I implemented something similar to [your change](https://github.com/awslabs/aws-codebuild-jenkins-plugin/commit/69244e36be3cc363b41c2ba0fb2c45cd840893fc) from a while back. This will be released today in version 0.36, so let us know once you've verified it works for your use case.
